sql: panic on INSERT: runtime error: index out of range

```
CREATE DATABASE d;
SET DATABASE = d;
CREATE TABLE ident ();
INSERT INTO ident AS TIME VALUES ( TRUE , DEFAULT ) ORDER BY TRUE , TRUE DESC OFFSET TRUE LIMIT TRUE;
```

Panics the server:

```
panic: runtime error: index out of range

goroutine 229 [running]:
panic(0x18a7020, 0xc82000e130)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/sql.(*planner).fillDefaults(0xc820394890, 0xc8202e6a30, 0x1, 0x1, 0xc82000b140, 0x1, 0x1, 0xc8202a0a50, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/insert.go:391 +0xe2e
github.com/cockroachdb/cockroach/sql.(*planner).Insert(0xc820394890, 0xc8202a0a50, 0x0, 0x0, 0x0, 0xc8202a2f01, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/insert.go:122 +0x7a8
github.com/cockroachdb/cockroach/sql.(*planner).newPlan(0xc820394890, 0x7fc1bb66d3e8, 0xc8202a0a50, 0x0, 0x0, 0x0, 0x25f94601, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/plan.go:260 +0x1535
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc820394890, 0x7fc1bb66d3e8, 0xc8202a0a50, 0x83fb01, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/plan.go:212 +0x83
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc8200a6000, 0x7fc1bb66d3e8, 0xc8202a0a50, 0xc820394890, 0xc8202a0a01, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:1045 +0x93
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc8200a6000, 0x7fc1bb66d3e8, 0xc8202a0a50, 0xc820394890, 0x101, 0xc820394828, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:943 +0x439
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc8200a6000, 0xc820582730, 0x1, 0x1, 0xc820394890, 0xc820394828, 0x1fe0101, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:691 +0x5da
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc8200a6000, 0xc820394890, 0x1, 0xc820394828, 0xc820828580, 0xc820582730, 0x1, 0x1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:619 +0x18d
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func2(0xc820359e40, 0xc820828580, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:511 +0x470
github.com/cockroachdb/cockroach/internal/client.(*Txn).Exec(0xc820359e40, 0xc820450101, 0x1461fececbe63048, 0x0, 0xc820459380, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:548 +0x2eb
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc8200a6000, 0x7fc1bc800908, 0xc820583f80, 0xc820394800, 0xc8207ce02d, 0xa0, 0x0, 0x0, 0x0, 0xc8207ce000)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:516 +0xba7
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8200a6000, 0x7fc1bc800908, 0xc820583f80, 0xc820394800, 0xc8207ce02d, 0xa0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:382 +0xf6
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc8204c2c00, 0x7fc1bc800908, 0xc820583f80, 0xc8207ce02d, 0xa0, 0x0, 0x0, 0x0, 0x0, 0x7fc1bc882201, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:640 +0x98
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc8204c2c00, 0x7fc1bc800908, 0xc820583f80, 0xc8204c2c28, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:320 +0xe8
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc8204c2c00, 0xc82087e720, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:275 +0x100c
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc8201cd740, 0x7fc1bc8c2800, 0xc8200b4300, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:229 +0x98f
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x7fc1bc8c20d0, 0xc820400000)
	/go/src/github.com/cockroachdb/cockroach/server/server.go:370 +0x42
github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1(0xc8201341e0, 0x7fc1bc8c20d0, 0xc820400000, 0xc8204f8010)
	/go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:131 +0x62
created by github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith
	/go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:133 +0x333
```

Found with the random syntax generator.

--- a/sql/descriptor_mutation_test.go
+++ b/sql/descriptor_mutation_test.go
@@ -252,7 +252,7 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR, i CHAR DEFAULT 'i');
 			t.Fatal(err)
 		}
 		// Repeating the same without specifying the columns results in a different error.
-		if _, err := sqlDB.Exec(`INSERT INTO t.test VALUES ('b', 'y', 'i')`); !testutils.IsError(err, "INSERT has more expressions than target columns: 3/2") {
+		if _, err := sqlDB.Exec(`INSERT INTO t.test VALUES ('b', 'y', 'i')`); !testutils.IsError(err, "INSERT error: table t.test has 2 columns but 3 values were supplied") {
 			t.Fatal(err)
 		}
 		// Make column "i" live so that it is read.

--- a/sql/insert.go
+++ b/sql/insert.go
@@ -139,7 +139,7 @@ func (p *planner) Insert(
 	}
 
 	if expressions := len(rows.Columns()); expressions > numInputColumns {
-		return nil, fmt.Errorf("INSERT has more expressions than target columns: %d/%d", expressions, numInputColumns)
+		return nil, fmt.Errorf("INSERT error: table %s has %d columns but %d values were supplied", n.Table, numInputColumns, expressions)
 	}
 
 	fkTables := TablesNeededForFKs(*en.tableDesc, CheckInserts)
@@ -385,7 +385,10 @@ func (p *planner) fillDefaults(defaultExprs []parser.TypedExpr,
 						&parser.Tuple{Exprs: append([]parser.Expr(nil), tuple.Exprs...)}
 					tupleCopied = true
 				}
-				if defaultExprs == nil {
+				if defaultExprs == nil || eIdx >= len(defaultExprs) {
+					// The case where eIdx is too large for defaultExprs will be
+					// transformed into an error by the check on the number of
+					// columns in Insert().
 					ret.Tuples[tIdx].Exprs[eIdx] = parser.DNull
 				} else {
 					ret.Tuples[tIdx].Exprs[eIdx] = defaultExprs[eIdx]

--- a/sql/testdata/insert
+++ b/sql/testdata/insert
@@ -294,7 +294,7 @@ NULL a
 statement ok
 CREATE TABLE boom (id INT PRIMARY KEY)
 
-statement error INSERT has more expressions than target columns: \d+/\d+
+statement error INSERT error: table boom has \d+ columns but \d+ values were supplied
 INSERT INTO boom VALUES(1,2)
 
 statement ok
@@ -435,3 +435,9 @@ CREATE TABLE blindcput (
 # many in one batch to trigger it.
 statement error duplicate key value \(x\)=\(1\) violates unique constraint "primary"
 INSERT INTO blindcput values (1, 1), (2, 2), (3, 3), (4, 4), (1, 5);
+
+statement ok
+CREATE TABLE nocols()
+
+statement error INSERT error: table nocols has \d+ columns but \d+ values were supplied
+INSERT INTO nocols VALUES (true, default)

